modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/client_http_invoker.ftl

```diff
@@ -241,6 +242,38 @@ public class HttpInvoker {
 		return fileName;
 	}
 
+	private HttpURLConnection _getHttpURLConnection(
+			HttpMethod httpMethod, String urlString)
+		throws IOException {
+
+		URL url = new URL(urlString);
+
+		HttpURLConnection httpURLConnection =
+			(HttpURLConnection)url.openConnection();
+
+		try {
+			HttpURLConnection setMethodHttpURLConnection = httpURLConnection;
+
+			if (Objects.equals(url.getProtocol(), "https")) {
+				Class<?> clazz = httpURLConnection.getClass();
+
+				Field field = clazz.getDeclaredField("delegate");
```

Comment:
Hey @drewbrokke, is this `delegate` field always present? Could it not exist if run in specific JVMs?

REST Builder is published globally for all versions of Liferay and must be backwards compatible with all of them. This means someone using 7.2 could run it tomorrow and get this change in their clients.

Reply:
I discussed this with @shuyangzhou yesterday, I had this question for him as well. He indicated that the internal structure of these JDK classes does not change often, if at all.

@shuyangzhou is there an automated way to run the included test case across the different supported JVMs? Just manually?

Reply:
Would it make sense to check if the field does not exist to just continue instead of failing?

Reply:
I don't think CI have the ability to spread the test on all supported jdks. You need to check it manually.

Reply:
When the internal structure changed, you will want a loud alarm to get your attention to fix it right away. "just continue instead of failing" means we will be sending out the request with a wrong method type, it will fail anyway, but in a very hidden style, it will cost more effort to find out what went wrong.

Reply:
Ok then if you're sure go for it. In any case I would recommend testing in different JVMs just in case.

Reply:
@shuyangzhou do we expect the client to add the `--add-opens` args when using our client in JDK 17 & 21? Is this documented anywhere? I think this client would already not work without them right now, even before this change.

But with this change, I had to include an additional "opens" statement for the `sun.net.www.protocol.https` package/module (not sure which) in order to get it to work with 17 & 21 (beyond the set we use [when starting tomcat](https://github.com/liferay/liferay-portal/blob/master/app.server.properties?plain=1#L29)).

Is there an expectation for the customer using our client libs to have to know/manage these things?

Reply:
@4lejandrito to answer your original question: Reflection access issues notwithstanding, it looks like the field is present (and the fix works) from Java 8 all the way through 21.

Reply:
We have added many --add-opens for tomcat for higher JDK. But I can not say anything about this "client", I have no idea how it is meant to be used.

Reply:
@shuyangzhou It's intended to be used by developers of custom apps (e.g. an external microservice or client extension). I tested in that scenario and indeed, I had to set the `-add-opens` JVM flags in order to use the client on JDK 17 & 21.

This change introduces _another_ layer of that, but it is already present with our currently released client libs. IMO this is not necessarily a much better developer experience than having an external dependency on the Apache HTTP client or something similar, but if we are going to put effort into migrating to an HTTP client, it seems better to focus on the JDK11+ client rather than Apache.

@4lejandrito do you know if we have any documentation that refers to the client libs? If so, could we consider including a section detailing the JVM args needed to run the client on higher JVMs?

Reply:
In any case, since the need for adding the JVM args is not _new_ with this change, then I think it's ok to proceed forward with it as-is.

Any objections, @4lejandrito?

Reply:
Go for it. I will check the docs and get back to you.

Reply:
Thanks @4lejandrito!

Reply:
Thanks for your review, @shuyangzhou!

Reply:
Hey @drewbrokke I believe we do not have official java client docs.